### PR TITLE
[Run] Introducing Run Launchers

### PR DESCRIPTION
**Pull request by alonmr**
As part of the separation of the server and client code, we introduce the launcher component which separates the different Runtime flows.
https://jira.iguazeng.com/browse/ML-2806

### BaseLauncher 🚀 
Abstract class for managing and running functions in different contexts.
This class is designed to encapsulate the logic of running a function in different contexts.
i.e. running a function locally, remotely or in a server.
Each context will have its own implementation of the abstract methods while the common logic resides in this class.

### ServerSideLauncher 🖥️ 
Contains the server-specific logic.
Triggered by submitting a job / or scheduling which starts runs from the server side by launching the functions (creates k8s pods etc.)

### ClientBaseLauncher 💻 
Abstract class for the client side launchers common logic.

### ClientLocalLauncher 📍 
ClientLocalLauncher is a launcher that runs the job locally.
Either on the user's machine (_is_run_local is True) or on a remote machine (_is_run_local is False).

### ClientRemoteLauncher 🌐 
ClientRemoteLauncher submits the run job to the server.

### LauncherFactory 🏭 
Creates the appropriate launcher for the specified run.
ServerSideLauncher - if running as API.
ClientRemoteLauncher - if run is remote and local was not specified.
ClientLocalLauncher - if run is not remote or local was specified.

### CODE REVIEW

File: mlrun/launcher/local.py [hunk 1]
```diff
@@ -0,0 +1,262 @@
...
+
+class ClientLocalLauncher(mlrun.launcher.client.ClientBaseLauncher):
+    """
+    ClientLocalLauncher is a launcher that runs the job locally.
+    Either on the user's machine (_is_run_local is True) or on a remote machine (_is_run_local is False).
+    """
+
+    def __init__(self, local: bool):
+        super().__init__()
+        self._is_run_local = local
```
Comment (liranbg) [inline mlrun/launcher/local.py:36]: Local on local client? Please elaborate or add docstirbg

**Review by liranbg — COMMENTED** [review]
Small

**Review by liranbg — APPROVED** [review]

File: mlrun/launcher/factory.py [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+import mlrun.config
+import mlrun.errors
+import mlrun.launcher.base
+import mlrun.launcher.local
+import mlrun.launcher.remote
+
+
+class LauncherFactory(object):
+    @staticmethod
+    def create_launcher(
```
Comment (theSaarco) [inline mlrun/launcher/factory.py:23]: Is there any advantage to creating a new launcher every time this is called? Aren't the launchers basically singletons in nature? Meaning, there shouldn't be more than 1 server-side launcher, 1 client-local launcher etc.?

Comment (alonmr) [reply mlrun/launcher/factory.py:23]: I will add a follow-up PR to change the classes to static classes / singletons

**Review by assaf758 — APPROVED** [review]

File: mlrun/launcher/base.py [hunk 3]
```diff
@@ -0,0 +1,391 @@
...
+                    run.spec.output_path = (
+                        mlrun.pipeline_context.project.spec.artifact_path
+                        or mlrun.pipeline_context.workflow_artifact_path
+                    )
+
+                if not run.spec.output_path and runtime._get_db():
+                    try:
+                        # not passing or loading the DB before the enrichment on purpose, because we want to enrich the
+                        # spec first as get_db() depends on it
+                        project = runtime._get_db().get_project(run.metadata.project)
```
Comment (Tankilevitch) [inline mlrun/launcher/base.py:292]: why not use self.db()?

File: mlrun/launcher/base.py [hunk 4]
```diff
@@ -0,0 +1,391 @@
...
+                    mlrun.pipeline_context.project
+                    and run.metadata.project
+                    == mlrun.pipeline_context.project.metadata.name
+                ):
+                    run.spec.output_path = (
+                        mlrun.pipeline_context.project.spec.artifact_path
+                        or mlrun.pipeline_context.workflow_artifact_path
+                    )
+
+                if not run.spec.output_path and runtime._get_db():
```
Comment (Tankilevitch) [inline mlrun/launcher/base.py:288]: is the `_get_db` is still needed?

**Review by Tankilevitch — CHANGES_REQUESTED** [review]
Looks good overall, there is still mix use of the `get_db` there is also one configured for each launcher and also one being set in the runtime. lets align that

**Pull request merged**